base_report_to_printer/models/ir_actions_report.py

```diff
@@ -0,0 +1,138 @@
+        company_dependent=True,
+    )
+    printing_printer_id = fields.Many2one(
+        comodel_name='printing.printer',
+        string='Default Printer'
+    )
+    printing_action_ids = fields.One2many(
+        comodel_name='printing.report.xml.action',
+        inverse_name='report_id',
+        string='Actions',
+        help='This field allows configuring action and printer on a per '
+             'user basis'
+    )
+
+    @api.model
+    def print_action_for_report_name(self, report_name):
+        """ Returns if the action is a direct print or pdf
+
+        Called from js
+        """
+        report = self._get_report_from_name(report_name)
+        if not report:
+            return {}
+        result = report.behaviour()[report]
+        serializable_result = {
+            'action': result['action'],
+            'printer_name': result['printer'].name,
+        }
+        return serializable_result
+
+    @api.multi
+    def behaviour(self):
+        result = {}
+        printer_obj = self.env['printing.printer']
+        printing_act_obj = self.env['printing.report.xml.action']
+        # Set hardcoded default action
+        default_action = 'client'
+        # Retrieve system wide printer
+        default_printer = printer_obj.get_default()
+
+        # Retrieve user default values
+        user = self.env.user
+        if user.printing_action:
+            default_action = user.printing_action
+        if user.printing_printer_id:
+            default_printer = user.printing_printer_id
+
+        for report in self:
+            action = default_action
+            printer = default_printer
+
+            # Retrieve report default values
+            report_action = report.property_printing_action_id
+            if report_action and report_action.action_type != 'user_default':
+                action = report_action.action_type
+            if report.printing_printer_id:
+                printer = report.printing_printer_id
+
+            # Retrieve report-user specific values
+            print_action = printing_act_obj.search([
+                ('report_id', '=', report.id),
+                ('user_id', '=', self.env.uid),
+                ('action', '!=', 'user_default'),
+            ], limit=1)
+            if print_action:
+                user_action = print_action.behaviour()
+                action = user_action['action']
+                if user_action['printer']:
+                    printer = user_action['printer']
+
+            result[report] = {
+                'action': action,
+                'printer': printer,
+            }
+        return result
+
+    @api.multi
+    def print_document(self, record_ids, data=None):
+        """ Print a document, do not return the document file """
+        document = self.with_context(
+            must_skip_send_to_printer=True).render_qweb_pdf(
+                record_ids, data=data)
+        behaviour = self.behaviour()[self]
```

Comment:
@sylvain-garancher I think I've noticed what may be a bug that is never triggered.  Note here that we call behaviour, and then use the self recordset as the index.  Now I can't think of a case where it isn't a singleton, but notice in `def behaviour` we iterate over self, so in theory there is a chance self is not in the result.  Indeed all 3 requests to that function originate from `print_document`, I just wonder if we are better to remove the loop in behaviour, use ensure_one() return a dict and remove the indexing in the 3 calls.  Would be a bit cleaner.

Reply:
Oops, this is an error I added during the migration, when I merged the `report` and `ir.actions.report.xml` models methods.
The original method was in the `report` model, and searched for a report, then used the report's id: https://github.com/OCA/report-print-send/blob/10.0/base_report_to_printer/models/report.py#L11-L23

I fully agree to add `self.ensure_one()` in this method. This is indeed always a singleton, because it was a single `report_name` argument before.

Reply:
@sylvain-garancher you might want to check my refactor at #112  which does this and completely reimplements some of the quirks around printer options which blocked me doing #111.  Its all rebased off this PR plus #110 .  It's just I'm not sure where to go, you've done all this port work, I've done a load of work on top of your port.  I kind of feel like the PR's should be reviewed on their own merits, but then merged together.

I have one more bug/design issue to consider.  If you create a user on 11.0 enterprise and do not give them the printing user security group you get a traceback on logging in.  

However, I wonder if there is really any point to this security group anyway and we don't just use base.group_user (or whatever it is now), or at least make it an implied group of that group.  

If I think about it, we can't stop a user printing a report (to screen) and clicking ctrl-p to print or emailing it.  In terms of accessing a report, well reports already have security.  It just feels like the group isn't really doing anything of value, and is very easy to miss when creating users.  

I struggle to think of a use case where a user is given a device to access an ERP, but not to use a printer.  If anything the security detail might be at the printer or server level.